Fix locale in contextcontroller

MelvinAchterhuis opened the pull request:
### 1. Why is this change necessary?

Currently when using a custom `config/routes/routes.yaml` the redirect after a language switch doesn't work correctly because the '_locale' that is submitted in the form is the locale of the current language and not the locale of the new language.

### 2. What does this change do, exactly?

Adds a hidden input field

`<input type="hidden" name="languageCode_{{ language.id }}" value="{{ language.translationCode.code }}">`

that is used to grab the language code from the new language and adds it as a redirect parameter

### 3. Describe each step to reproduce the issue or behaviour.

Add custom routes.yaml:

```
frontend.wishlist.page:
    path:
        en-GB: '/wishlist'
        de-DE: '/merkliste'
    controller: 'Shopware\Storefront\Controller\WishlistController::index'
    methods: ['GET']
    defaults:
        _noStore: true
        _routeScope: ['storefront']
    options:
        seo: false
frontend.account.home.page:
    path:
        en-GB: '/account'
        de-DE: '/konto'
    controller: 'Shopware\Storefront\Controller\AccountProfileController::index'
    methods: ['GET']
    defaults:
        _noStore: true
        _loginRequired: true
        _routeScope: ['storefront']
```

Go to wishlist or account home page and switch language. Localized path isn't respected because of this piece of code that adds the current locale:

```
{% for key, value in app.request.attributes.get('_route_params') %}
        <input name="redirectParameters[{{ key }}]" type="hidden" value="{{ value }}">
{% endfor %}
```

For example, when switching on /account from English to German the en-GB locale is in the route_params (current locale): 

`<input name="redirectParameters[_locale]" type="hidden" value="en-GB">`

This results in a redirect to /de/account instead of /de/konto 

Before change (with routes.yaml):
https://github.com/shopware/shopware/assets/26538915/0ab41131-4987-44a1-916a-76ef4059149d

After change (with routes.yaml):
https://github.com/shopware/shopware/assets/26538915/9c522f6a-1806-40a9-b16d-56d40f622c4a

### 4. Please link to the relevant issues (if any).
N/A

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

MelvinAchterhuis committed:
draft: fix locale in contextcontroller

MelvinAchterhuis committed:
draft: fix locale in contextcontroller

MelvinAchterhuis commented:
I've simplified the code by adding a hidden input with a combination of languageId + languageCode so we can grab that in the context controller and pass it as a locale param.

Not sure how to test this though, because Shopware doesn't use localised routes 😄

MelvinAchterhuis commented:
@keulinho , can this be reviewed? Thx 😄

ssltg commented:
@MelvinAchterhuis thank you for your distribution. It looks good so far. Can you please add a changelog?

Regards

MelvinAchterhuis commented:
> @MelvinAchterhuis thank you for your distribution. It looks good so far. Can you please add a changelog?
> 
> Regards

Yes, will do this week + fix conflicts.

MelvinAchterhuis committed:
add hidden input after rebase

add hidden input after rebase

MelvinAchterhuis committed:
Add changelog

MelvinAchterhuis commented:
@ssltg , I've fixed the conflicts and added a changelog. Should we mention this feature in the Shopware documentation or are we going to assume that people will read the Symfony documentation when they want to localise routes?

ssltg commented:
@MelvinAchterhuis Thak you 👍 
At least we should add a mention about it in the docs with a link to the symfony docs. Maybe a brief introduction would be great as well.

Regards

mitelg commented:
thanks for your contribution @MelvinAchterhuis 👍🎉💙
